pkg/machineproviders/providers/openshift/machine/v1beta1/provider.go

```diff
@@ -128,16 +170,296 @@ type openshiftMachineProvider struct {
+		if fd.Equal(failureDomain) {
+			return i, true
+		}
+	}
+
+	return 0, false
+}
+
+// isMachinesNamesFormatCorrect determines whether all provided machines names have correct format.
+func isMachinesNamesFormatCorrect(machines []machinev1beta1.Machine) bool {
+	for _, machine := range machines {
+		// Get a substring after the last occurrence of "-" to find the machine index suffix.
+		// Then compare it with the regexp to check if it is a number.
+		if !digitCheck.MatchString(machine.Name[strings.LastIndex(machine.Name, "-")+1:]) {
+			return false
+		}
+	}
+
+	return true
+}
+
+// getMachineRef returns returns machine object reference for the given machine.
+func getMachineRef(machine machinev1beta1.Machine) *machineproviders.ObjectRef {
+	return &machineproviders.ObjectRef{
+		ObjectMeta:           machine.ObjectMeta,
+		GroupVersionResource: machinev1beta1.GroupVersion.WithResource("machines"),
+	}
+}
+
+// getNodeRef returns node object reference for the given machine.
+func getNodeRef(machine machinev1beta1.Machine) *machineproviders.ObjectRef {
+	if machine.Status.NodeRef == nil {
+		return nil
+	}
+
+	return &machineproviders.ObjectRef{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: machine.Status.NodeRef.Name,
+		},
+		GroupVersionResource: corev1.SchemeGroupVersion.WithResource("nodes"),
+	}
+}
+
+// getMachineErrorMessage returns error message for the given machine.
+func getMachineErrorMessage(machine machinev1beta1.Machine) string {
+	if machine.Status.ErrorMessage == nil {
+		return ""
+	}
+
+	return *machine.Status.ErrorMessage
+}
+
+// getMachineReadiness returns if machine is ready.
+func getMachineReadiness(machine machinev1beta1.Machine) bool {
+	return machine.Status.Phase != nil && *machine.Status.Phase == runningPhase
 }
 
 // CreateMachine creates a new Machine from the template provider config based on the
 // failure domain index provided.
 func (m *openshiftMachineProvider) CreateMachine(ctx context.Context, logger logr.Logger, index int32) error {
+	machineName, err := m.getMachineName(index)
+	if err != nil {
+		return fmt.Errorf("could not generate machine name: %w", err)
+	}
+
+	cpms := &machinev1.ControlPlaneMachineSet{
+		ObjectMeta: m.ownerMetadata,
+	}
+
+	machine := &machinev1beta1.Machine{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        machineName,
+			Namespace:   m.namespace,
+			Annotations: m.machineTemplate.ObjectMeta.Annotations,
+			Labels:      m.machineTemplate.ObjectMeta.Labels,
+		},
+		Spec: m.machineTemplate.Spec,
+	}
+
+	pc, err := m.providerConfig.InjectFailureDomain(m.indexToFailureDomain[index])
+	if err != nil {
+		return fmt.Errorf("cannot inject failure domain in the provider config: %w", err)
+	}
+
+	rawConfig, err := pc.RawConfig()
+	if err != nil {
+		return fmt.Errorf("cannot fetch raw config from provider config: %w", err)
+	}
+
+	machine.Spec.ProviderSpec.Value.Raw = rawConfig
+
+	err = controllerutil.SetControllerReference(cpms, machine, m.machineAPIScheme)
+	if err != nil {
```

Comment:
Inline this and don't shadow please
```suggestion
	if err = controllerutil.SetControllerReference(cpms, machine, m.machineAPIScheme); err != nil {
```